pip/utils/temp_dir.py

```diff
@@ -6,37 +6,40 @@
 from pip.utils import rmtree
 
 
-class BuildDirectory(object):
+class TempDirectory(object):
+    """Helper class that owns and cleans up a temporary directory.
+    """
 
-    def __init__(self, name=None, delete=None):
-        # If we were not given an explicit directory, and we were not given an
-        # explicit delete option, then we'll default to deleting.
-        if name is None and delete is None:
-            delete = True
+    def __init__(self, path=None, delete=None, kind="temp"):
+        super(TempDirectory, self).__init__()
 
-        if name is None:
+        self.path = path
+        self.delete = delete
+        self.kind = kind
+
+    def __repr__(self):
+        return "<{} {!r}>".format(self.__class__.__name__, self.path)
+
+    def __enter__(self):
+        if self.path is None:
             # We realpath here because some systems have their default tmpdir
             # symlinked to another directory.  This tends to confuse build
             # scripts, so we canonicalize the path by traversing potential
             # symlinks here.
-            name = os.path.realpath(tempfile.mkdtemp(prefix="pip-build-"))
+            self.path = os.path.realpath(
+                tempfile.mkdtemp(prefix="pip-", suffix="-" + self.kind)
+            )
             # If we were not given an explicit directory, and we were not given
             # an explicit delete option, then we'll default to deleting.
-            if delete is None:
-                delete = True
-
-        self.name = name
-        self.delete = delete
-
-    def __repr__(self):
-        return "<{} {!r}>".format(self.__class__.__name__, self.name)
+            if self.delete is None:
+                self.delete = True
 
-    def __enter__(self):
-        return self.name
+        return self
 
     def __exit__(self, exc, value, tb):
-        self.cleanup()
+        if self.delete:
+            self.cleanup()
 
     def cleanup(self):
-        if self.delete:
-            rmtree(self.name)
+        if os.path.exists(self.path):
+            rmtree(self.path)
```

Comment:
I'm not sure I see the value of having a separate ```cleanup``` method. We don't use it in the existing code, and I'm struggling to see a good use case for it. If it *is* useful, maybe we should document the expected pattern of use, showing how people should use it.